text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+example we need an identifier to finish the `.` expression, a `;` to terminate
+the let statement, and a `}` to terminate the `main` function).
+
+A solution to the first problem is replacing invalid names with some magic
+identifier, and ignoring errors involving that identifier. @sanxiyn implemented
+something like the second feature in a [PR](https://github.com/rust-
+lang/rust/pull/21323). His approach was to take a command line argument for
+where to 'complete at' and to treat that as the magic identifier. An alternate
+approach would be to use a keyword or distinguished identifier which the IDE
+could insert (based on the caret position), or to fallback to the magic
+identifier whenever there is a name resolution error.
+
+Similarly during type checking, if we find a mismatched or unknown type, we
+should try to continue type checking with the information available so as to
+still be able to provide code completion information. We already do this to some
+extent with `TyErr`, but we should do better.
+
+For the second issue, the problem is where to start parsing again and how many
+'open' items should be terminated. This is closely related to error recovery in
+parsers, which is a well-developed are of research with a long history, and
+which I won't attempt to summarise here. As far as I can see, there are two
+major differences since we are doing this in the IDE context: we know the extent
+of edited code (the span of changes we are passing to the quick-check compiler)
+and the previous state of the edited code, and we can likely assume that even in
+new code, braces and parentheses are likely to be paired (since an IDE will
+insert closing braces, etc.). Assuming that we keep the state of the code the
+last time it parsed completely, we can expand the edited span to cover an entire
+expression (or other item) and thus we know exactly where to start re-parsing.
+In the case where we are writing new code, we can just close all 'open' items.
+
+Being able to generate more errors before stopping would be an advantage for the
+compiler in any case. However, we probably do not want to use these mechanisms
+under normal compilation, only when performing a quick-check from the IDE.
+
+
+
+## Error format
+
+Currently the compiler generates text error messages. I propose that we add a
+mechanism to the compiler to support different formats for error messages. We
+already structure our error messages to some extent (separating the span
+information, the message, and the error code). Rather than turning these
+components into text in a fairly ad hoc manner, we should preserve that
+structure, and some central error handler should convert into a chosen format.
+We should support the current text format, JSON (or some other structured
+format) for tools to use, and HTML for rich error messages (this is somewhat
+orthogonal to this RFC, but has been discussed in the past as a desirable
+feature).
```

Comment:
Could you give an example of how an HTML error message could be useful? Or give a link to a discussion on the subject?